feat: Rewrite the CI

pmnoxx opened the pull request:
I'm adding `.github/workflows/rust.yml`, that should configure the github pipeline for borsh.

matklad commented:
Thanks for working on this!

* let's remove travis.yml completely
* let's use actions-rs to install a specific version of Rust, instead of relying on whatever is packaged with ubuntu. I think you'll need something like this: 

```
    - uses: actions-rs/toolchain@v1
      with:
        toolchain: stable
```

* let's remove separate build job, running the tests should be enough
* let's move coverage into a separate coverage job (I am not sure if coverage is working here, maybe it's better to remove it altogether, but I'd rather not do this in this PR)

matklad commented:
Actually, I'd also maybe remove the cache step:

* it takes quite a bit of yaml to implement
* borsh build should be fast enough for caching to not matter
* what we do here is suboptimal anyway, if caching does matter, Swatinem/rust-cache. should be used https://matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching

austinabell commented:
> Actually, I'd also maybe remove the cache step:
> 
> * it takes quite a bit of yaml to implement
> * borsh build should be fast enough for caching to not matter
> * what we do here is suboptimal anyway, if caching does matter, Swatinem/rust-cache. should be used [matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching](https://matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching)

https://github.com/Swatinem/rust-cache is a one-line option that seems to have benefits for a lot of libs, perhaps this is a third option aside from manual or none

pmnoxx commented:
> Actually, I'd also maybe remove the cache step:
> 
> * it takes quite a bit of yaml to implement
> * borsh build should be fast enough for caching to not matter
> * what we do here is suboptimal anyway, if caching does matter, Swatinem/rust-cache. should be used https://matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching

I originally added caching to make `cargo-deny` installation a bit faster. It was taking around 7m without cache, and 1m with cache. However, I removed `cargo-deny`, so the cache is no longer needed.

Besides, it's best to first add CI as simple as possible, and then do bench-marking, and decide whenever it's needed or not.
I removed the cache for now.

pmnoxx commented:
> * borsh build should be fast enough for caching to not matter

Tests take 3m58s without cache and 2m36s with cache. 
https://github.com/near/borsh-rs/actions/runs/1608140379 vs https://github.com/near/borsh-rs/actions/runs/1595981247

pmnoxx commented:
I

> > Actually, I'd also maybe remove the cache step:
> > 
> > * it takes quite a bit of yaml to implement
> > * borsh build should be fast enough for caching to not matter
> > * what we do here is suboptimal anyway, if caching does matter, Swatinem/rust-cache. should be used [matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching](https://matklad.github.io/2021/09/04/fast-rust-builds.html#ci-caching)
> 
> https://github.com/Swatinem/rust-cache is a one-line option that seems to have benefits for a lot of libs, perhaps this is a third option aside from manual or none

I tried using 
```
 - uses: actions-rs/toolchain@v1
      with:
        profile: minimal
        toolchain: stable
```.
However, 2nd run take 4m. So it seems it doesn't work.

frol commented on .github/workflows/rust.yml:
```diff
@@ -0,0 +1,76 @@
+  CARGO_TERM_COLOR: always
+
+jobs:
+  build:
+
+    runs-on: ubuntu-20.04
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build
+      run: RUSTFLAGS='-D warnings' cargo build --verbose --workspace
```
It is unclear what we want to achieve with this stage given that borsh-rs is a library crate

pmnoxx replied:
@frol This stage is for unit tests. Tests need to be build, and all their dependencies need to be downloaded from the internet. We can cut time of doing tests from 4m to 2m30s by using cache. 

Though, we don't ship many PR's, even if the `tests` takes 4m to finish, it's not that big of a deal.

miraclx replied:
If that's what you want, `cargo build` requires `--tests` to actually build the tests. You could also use `cargo test --no-run` instead.

pmnoxx replied:
> It is unclear what we want to achieve with this stage given that borsh-rs is a library crate

Oh, that's what you meant. Yes, we can remove cargo build.

frol reviewed (approved):
The motivation is unclear to me (e.g. why did we only ported part of the CI and keep travis around?), but I am fine with whatever CI rules we come up, the faster the CI the better for all the contributors.

pmnoxx commented:
> The motivation is unclear to me (e.g. why did we only ported part of the CI and keep travis around?), but I am fine with whatever CI rules we come up, the faster the CI the better for all the contributors.

@matklad I need to figure out how to remove Travis CI. Removing `.travis.yml` is enough enough.
